### ISSUE TITLE
Prevent `METRICS_PUSHGATEWAY_URL ` environment error in Debug builds

### ISSUE THREAD

**Issue by daveverwer** [SwiftPackageIndex/SwiftPackageIndex-Server#1182]
```
[ WARNING ] Environment variable not set: METRICS_PUSHGATEWAY_URL [component: server] Swift/ErrorType.swift:200: Fatal error: Error raised at top level: App.AppError.envVariableNotSet("METRICS_PUSHGATEWAY_URL")
```

**Comment by Sherlouk** [comment]
❤️  every time this error catches me out

**Comment by finestructure** [comment]
What's the dev use case where this warning actually shows up? I was just looking into this and realised I always run via Xcode and don't get this:

<img width="556" alt="CleanShot 2021-07-05 at 08 08 19@2x" src="https://user-images.githubusercontent.com/65520/124424960-2e6a7c80-dd68-11eb-9199-63dde8ae7b96.png">

Are you actually building a docker image and running it via compose? If so, my initial approach to detect a debug build might not work (I think the image is built in release mode). We also can't use `Environment.detect`, because `dev` is staging - so we have no local environment to suppress this on unless we introduce a new one and make sure it's set everywhere. I.e. it's easy to end up in the same situation as before if the env isn't correctly set.

**Comment by daveverwer** [comment]
I don't get it in Xcode anymore, but I certainly have in the past, and it's certainly not when building with docker. Strange!

@selichter and @Sherlouk what are we doing wrong?

**Comment by Sherlouk** [comment]
Aye, I think this must have been fixed by another PR because I'm not getting it when I build from `main`.

**Comment by finestructure** [comment]
I'll go ahead and close this as magically fixed 😬  Please re-open if it appears again!

**Comment by finestructure** [comment]
One thing to note about this one: this is a warning message and it does not crash for me:

```
❯ swift run Run ingest --limit 1
[0/0] Build complete!
[ INFO ] Ingesting (limit: 1) ... [component: ingest]
[ WARNING ] AppMetrics.push failed with error: connection reset (error set): Connection refused (errno: 61) [component: ingest]
```

I see it does/did for you, Dave:

```
[ WARNING ] Environment variable not set: METRICS_PUSHGATEWAY_URL [component: server]
Swift/ErrorType.swift:200: Fatal error: Error raised at top level: App.AppError.envVariableNotSet("METRICS_PUSHGATEWAY_URL")
make: *** [ingest] Trace/BPT trap: 5
```

That's something it definitely shouldn't do but I think this might be something that's now been fixed with the transition to async/await, where I made changes to the error handling. We might have let the exception escape before.

**Comment by finestructure** [comment]
Hold on - that's an `Connection refused (errno: 61)` error, which I got because I'd set 

```
METRICS_PUSHGATEWAY_URL=http://127.0.0.1:9091
```

in my `.env` without actually running the push service. If I remove that, I'm getting:

```
❯ swift run Run ingest --limit 1
[0/0] Build complete!
[ INFO ] Ingesting (limit: 1) ... [component: ingest]
~/P/S/spi-server on  main took 2s
```

whereas `ingest` logs

```
❯ env swift run Run analyze --limit 1
[0/0] Build complete!
[ INFO ] Analyzing (limit: 1) ... [component: analyze]
[ INFO ] Checkout directory: /Users/sas/Projects/SPI/spi-server/SPI-checkouts [component: analyze]
[ INFO ] cloning https://github.com/LebJe/TOMLKit.git to /Users/sas/Projects/SPI/spi-server/SPI-checkouts/github.com-lebje-tomlkit [component: analyze]
[ INFO ] listing tags for package https://github.com/LebJe/TOMLKit.git [component: analyze]
[ WARNING ] Twitter.postToFirehose failed: The operation couldn’t be completed. (App.Twitter.Error error 2.) [component: analyze]
[ WARNING ] Environment variable not set: METRICS_PUSHGATEWAY_URL [component: server]
Swift/ErrorType.swift:200: Fatal error: Error raised at top level: App.AppError.envVariableNotSet("METRICS_PUSHGATEWAY_URL")
fish: Job 1, 'env GITHUB_TOKEN=ghp_LZhZUdXlMF…' terminated by signal SIGTRAP (Trace or breakpoint trap)
~/P/S/spi-server on  main took 6s
```

So the async/await command (`ingest`) is swallowing the exception while the ELF one (`analyse`) isn't. I specifically dealt with this in 9d882451dac0d5653addb427a7ba31c980a0ec53 but I must have messed it up subsequently.

**Comment by finestructure** [comment]
Ok, I'd subsequently silenced the metrics errors but that was a mistake. The difference to ELF is that we can't let exceptions bubble out of the async command anymore (until there is a proper async `Command`). `run` as we have it in the new a/a commands can't be throwing for the moment.

We're not relying on the command status codes anywhere, so this is not an issue. All our monitoring is looking for `CRITICAL` and `ERROR` logs etc.

**Comment by finestructure** [comment]
This was fixed in #1522
